parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -1069,27 +1073,76 @@ pushGitRootBranch srcConn branch repo = runExceptT @GitError do
       chId <- fromMaybe err <$> Q.loadCausalHashIdByCausalHash h2
       Q.setNamespaceRoot chId
 
+    -- This function makes sure that the result of git status is valid.
+    -- Valid lines are any of:
+    --
+    --   ?? .unison/v2/unison.sqlite3 (initial commit to an empty repo)
+    --   M .unison/v2/unison.sqlite3  (updating an existing repo)
+    --   D .unison/v2/unison.sqlite3-wal (cleaning up the WAL from before bugfix)
+    --   D .unison/v2/unison.sqlite3-shm (ditto)
+    --
+    -- Invalid lines are like:
+    --
+    --   ?? .unison/v2/unison.sqlite3-wal
+    --
+    -- Which will only happen if the write-ahead log hasn't been
+    -- fully folded into the unison.sqlite3 file.
+    --
+    -- Returns `Just (hasDeleteWal, hasDeleteShm)` on success,
+    -- `Nothing` otherwise. hasDeleteWal means there's the line:
+    --   D .unison/v2/unison.sqlite3-wal
+    -- and hasDeleteShm is `True` if there's the line:
+    --   D .unison/v2/unison.sqlite3-shm
+    --
```

Comment:
I made this more complicated than we originally discussed, to gracefully handle the case where a push cleans up / deletes old WAL files. The implementation of this is kinda fragile if the format of `git status --short` changes but I think it's okay.

Reply:
Also somewhat reassuring is I had this function wrong in like every possible way (forgot to allow for the `unison.sqlite3` file to be created, forgot to allow the wal file to be deleted, forgot the `-uall` for pushing to an empty repo...) and got test failures or failures in some manual testing.